Models

aehyvari opened the pull request:
This PR suggests a partial redesign of the model system motivated by the need to produce models for EUF.

 - The substitution - rewriting part of the code has some changes resulting in more substitutions being tracked.
 - The substitution handling in model production is delegated to theory solvers
 - The Logic::TFun class is refactored into two standalone classes for template functions and template function signatures
 - The model printing complies to the smtlib2 standard
 - The UF models for functions are stored in a tree structure that avoids (hopefully) efficiently insertion of duplicate models

For now the models are tested for QF_LIA, QF_LRA, and QF_UF, but only on a relatively short, 10 s timeout running on unoptimised debug code so I can't rule out surprises.  Before merging it would make sense to do a bigger test in cluster.

There seems to be in general an overhead because of the substitution-rewriting changes.  However, in all three tested logics the number of instances solved with the new substitution is bigger than with the old substitution.

### QF_UF
[This plot](https://user-images.githubusercontent.com/8374962/117820523-ea2c9580-b26a-11eb-9687-e8435d156a26.png) shows the overhead in pre-processing only code (checking disabled).  There is a clear overhead.

It looks like the change has a positive effect though on the overall solving, although this is debatable.  [See here](https://user-images.githubusercontent.com/8374962/117820768-24963280-b26b-11eb-825d-5fa804812299.png) for the comparison, and in particular note the long-running SAT instances, the speedup in a few UNSAT instances.  The timeouts in master solved in uf-models might be a random effect, a result of non-equal timeouts, or a real speed-up with the new simplifications.

### QF_LIA
[Comparison in QF_LIA](https://user-images.githubusercontent.com/8374962/117994084-8f656d80-b340-11eb-9ec9-89c626d871e8.png) shows that the substitutions make a big difference.  Overall result is positive in terms of number of solved instance.

### QF_LRA
For linear real arithmetics [the results](https://user-images.githubusercontent.com/8374962/117995200-6b565c00-b341-11eb-803f-89fe0167130e.png) suggest a %-overhead.  I can't really explain this, since the pre-processing is just an initial pass and therefore should be a constant overhead.  I think I'm over-interpreting the plot.

aehyvari commented on src/logics/Theory.cc:
```diff
@@ -52,26 +52,27 @@ Theory::SubstitutionResult Theory::computeSubstitutions(const PTRef fla)
-                allsubsts.insert(key, newsubsts_vec[i].data);
+        auto newRootAndSubsts(Substitutor<PtAsgn>(getLogic(), newsubsts).rewrite(root));
+        PTRef new_root = newRootAndSubsts.first;
+        auto usedSubstitutions(std::move(newRootAndSubsts.second));
```
Here's essentially why I felt that I need to change the interface of the Substitutor.  The substitutions that are computed are used by the models (and indeed also by this very substitution loop).  It might be good to write some templates to avoid having to deal every time with the returned substitution map in places where it's not needed.

blishko commented on src/logics/Logic.h:
```diff
@@ -93,37 +51,33 @@ class Logic {
     Map<SRef,bool,SRefHash,Equal<SRef> >            ufsorts;
+    Map<SRef,PTRef,SRefHash,Equal<SRef>>            defaultValueForSort;
 
+    bool isKnownToUser(SymRef sr) const { return getSymName(sr)[0] != s_abstract_value_prefix[0]; }
```
I am wondering if `s_abstract_value_prefix` should not be just a single `char` instead of `const char*`.

aehyvari replied:
I'm not sure.  In the end it probably matters little.

blishko commented on src/logics/Logic.h:
```diff
@@ -93,37 +51,33 @@ class Logic {
-            defined_functions_names.last() = strdup(name);
-            defined_functions[defined_functions_names.last()] = templ;
+            defined_functions_names.emplace_back(name);
+            defined_functions[defined_functions_names.back()] = std::move(templ);
```
```suggestion
            defined_functions[name] = std::move(templ);
```

aehyvari replied:
Yes.

blishko commented on src/minisat/mtl/Map.h:
```diff
@@ -136,6 +136,12 @@ class Map {
     // PRECONDITION: the key must *NOT* exist in the map.
     void insert (const K& k, const D& d) { if (checkCap(size+1)) rehash(); _insert(k, d); size++; }
+    /**
+     * Check if k exists in the map.  If it exists, return h[k] i d.  If it does not, d remains unchanged
```
What does `return h[k] i d` mean?

aehyvari replied:
*in

blishko commented on src/models/Model.cc:
```diff
@@ -3,6 +3,30 @@
+           if (s.nargs() != templFun.getArgs().size_()) { std::cout << "Argument number doesn't match" << std::endl; return false; }
+           if (logic.getSortRef(sr) != templFun.getRetSort()) { std::cout << "Return sorts don't match" << std::endl; return false; }
+           for (auto i = 0; i < (int)s.nargs(); i++)
+               if (s[i] != logic.getSortRef(templFun.getArgs()[i])) { std::cout << "argument sort doesn't match" << std::endl; return false; }
```
Should these be `std::cerr` instead of `std::cout`?

aehyvari replied:
I'll just remove the prints.

blishko reviewed (commented):
I made a first pass with some small comments.
I have to look at the changes for the substitutions in more detail.
I am not very keen on having the templated version for both `PTRef and PTAsgn` and I don't like the templated `DefaultRewriteConfig` or that the `rewrite` method would return a pair. I hope there is another way to achieve what is needed.

aehyvari commented on src/bin/Interpret.cc:
```diff
@@ -709,18 +709,64 @@ void Interpret::getModel() {
+    SRef sortRef = logic->getSym(tr).rsort();
+    ss << "  (define-fun " << s << " () " << logic->getSortName(sortRef) << '\n';
+    char* val_string = logic->pp(val);
+    std::string val_s(val_string);
```
Looks like so.

aehyvari commented on src/logics/Logic.cc:
```diff
@@ -501,6 +515,9 @@ bool Logic::declare_sort_hook(SRef sr) {
     sortToIte.insert(sr, tr);
 
+    std::stringstream ss;
+    ss << Logic::s_abstract_value_prefix << "d" << sort_store.numSorts();
```
Ok.

aehyvari commented on src/logics/Logic.cc:
```diff
@@ -283,7 +283,17 @@ Logic::protectName(const char* name) const
+}
+
+char*
+Logic::protectName(const std::string& name) const
```
Yes.

aehyvari commented on regression/model/moreNonUFsIntroducedForEgraph.smt2:
```diff
@@ -0,0 +1,1001 @@
+(assert (= x574 (= x611 x550)))
+(check-sat)
+(get-model)
+(exit)
```
It could actually be ok to shorten this one.  There's another regression test now that's been more useful at catching problems.

aehyvari commented:
I rebased this on current master, rebased the couple of commits from @blishko's models branch on top of this and addressed some of the comments.  I'm running some bigger scale tests to see if things are starting to work...

blishko reviewed (commented):
Small batch of comments, I still have not checked how the `Egraph` computes the model and how this information gets to `Model`.

blishko commented on src/tsolvers/egraph/EgraphSolver.cc:
```diff
@@ -213,23 +215,131 @@ void Egraph::getConflict( bool deduction, vec<PtAsgn>& cnfl )
+                target_er = er;
+            }
+        } else {
+            assert(([&](PTRef tr) { ERef tmp; enode_store.peekERef(tr, tmp); return tmp == er; })(tr));
```
This is a nice trick :)
But I am not sure what it says. Should `tr` be in a the store or not? I think this can be expressed more clearly.

aehyvari replied:
I think it's just a consistency check on `enode_store`, that it's map from `PTRef` -> `ERef` is consistent with the other direction, `ERef -> PTRef`.

blishko commented on src/tsolvers/egraph/EgraphSolver.cc:
```diff
@@ -213,23 +215,131 @@ void Egraph::getConflict( bool deduction, vec<PtAsgn>& cnfl )
+        } else {
+            assert(([&](PTRef tr) { ERef tmp; enode_store.peekERef(tr, tmp); return tmp == er; })(tr));
+            assert(enode_store.peekERef(tr, target_er));
+            assert(target_er == er);
```
Actually, don't these two asserts say the same thing as the one with lambda?

aehyvari replied:
Yes, I agree.  The lambda looks a little nicer, at least it's everything in one place...

aehyvari commented on src/logics/Logic.cc:
```diff
@@ -1255,28 +1264,31 @@ lbool Logic::retrieveSubstitutions(const vec<PtAsgn>& facts, MapWithKeys<PTRef,P
-            PTRef oldVal = val.tr;
-            PTRef newVal = Substitutor(*this, substs).rewrite(oldVal);
+            PTRef & val = getPTRefFromAsgn<TAsgn>(substs[keys[i]]);
+            if (notChangedElems[i]) { continue; }
```
Isn't this a bug?  I'd say that it's not correct to rewrite the disabled substitutions.

aehyvari commented on src/logics/Logic.cc:
```diff
@@ -1285,6 +1297,9 @@ void Logic::substitutionsTransitiveClosure(MapWithKeys<PTRef, PtAsgn, PTRefHash>
 
+template void Logic::substitutionsTransitiveClosure<PtAsgn>(MapWithKeys<PTRef,PtAsgn,PTRefHash> &);
+template void Logic::substitutionsTransitiveClosure<PTRef>(MapWithKeys<PTRef,PTRef,PTRefHash> &);
+
```
The explicit template instantiations are needed here at least by clang on my mac.  I think it's got something to do with the way templates are created on one go, but I'm not sure if there's some differences how compilers do it.

aehyvari commented on src/logics/Theory.cc:
```diff
@@ -57,19 +56,25 @@ Theory::SubstitutionResult Theory::computeSubstitutions(const PTRef fla)
+                allsubsts.insert(key, target.tr);
             }
         }
+        getLogic().substitutionsTransitiveClosure(allsubsts);
```
It sounds to me that once outside the loop should be enough.
